node/src/components/contract_runtime.rs

```diff
@@ -202,362 +319,416 @@ where
+                        upgrade_config,
+                        responder,
+                    } => {
+                        trace!(?upgrade_config, "upgrade");
+                        let engine_state = Arc::clone(&self.engine_state);
+                        let metrics = Arc::clone(&self.metrics);
+                        async move {
+                            let correlation_id = CorrelationId::new();
+                            let start = Instant::now();
+                            let result =
+                                engine_state.commit_upgrade(correlation_id, *upgrade_config);
+                            metrics
+                                .commit_upgrade
+                                .observe(start.elapsed().as_secs_f64());
+                            trace!(?result, "upgrade result");
+                            responder.respond(result).await
+                        }
+                        .ignore()
+                    }
+                    ContractRuntimeRequest::Query {
+                        query_request,
+                        responder,
+                    } => {
+                        trace!(?query_request, "query");
+                        let engine_state = Arc::clone(&self.engine_state);
+                        let metrics = Arc::clone(&self.metrics);
+                        async move {
+                            let correlation_id = CorrelationId::new();
+                            let start = Instant::now();
+                            let result = engine_state.run_query(correlation_id, query_request);
+                            metrics.run_query.observe(start.elapsed().as_secs_f64());
+                            trace!(?result, "query result");
+                            responder.respond(result).await
+                        }
+                        .ignore()
+                    }
+                    ContractRuntimeRequest::GetBalance {
+                        balance_request,
+                        responder,
+                    } => {
+                        trace!(?balance_request, "balance");
+                        let engine_state = Arc::clone(&self.engine_state);
+                        let metrics = Arc::clone(&self.metrics);
+                        async move {
+                            let correlation_id = CorrelationId::new();
+                            let start = Instant::now();
+                            let result = engine_state.get_purse_balance(
+                                correlation_id,
+                                balance_request.state_hash(),
+                                balance_request.purse_uref(),
+                            );
+                            metrics.get_balance.observe(start.elapsed().as_secs_f64());
+                            trace!(?result, "balance result");
+                            responder.respond(result).await
+                        }
+                        .ignore()
```

Comment:
Chatting with @dwerner , these spawn blockings were removed and we did not see any regressions.